src/mlpack/methods/neat/genome_impl.hpp

```diff
@@ -0,0 +1,523 @@
+  }
+  size_t sourceID = connectionGeneList[i].Source();
+  size_t targetID = connectionGeneList[i].Target();
+  size_t newNodeID = nextNodeID++;
+  size_t innovID1, innovID2;
+
+  // Check if these mutations have been made. Else, add them to the buffer.
+  std::pair<size_t, size_t> key1 = std::make_pair(sourceID, newNodeID);
+  std::pair<size_t, size_t> key2 = std::make_pair(newNodeID, targetID);
+
+  if (mutationBuffer.find(key1) == mutationBuffer.end())
+  {
+    innovID1 = nextInnovID++;
+    mutationBuffer[key1] = innovID1;
+  }
+  else
+    innovID1 = mutationBuffer[key1];
+
+  if (mutationBuffer.find(key2) == mutationBuffer.end())
+  {
+    innovID2 = nextInnovID++;
+    mutationBuffer[key2] = innovID2;
+  }
+  else
+    innovID2 = mutationBuffer[key2];
+
+  // Add the first connection to the containers.
+  directedGraph[sourceID].emplace(newNodeID, ConnectionGene(innovID1, 1,
+      sourceID, newNodeID));
+  connectionGeneList.emplace_back(ConnectionGene(innovID1, 1,
+      sourceID, newNodeID));
+
+  // Add the second connection to the containers.
+  connectionGeneList.emplace_back(ConnectionGene(innovID2, 1,
+      newNodeID, targetID));
+  directedGraph.emplace(std::piecewise_construct,
+                        std::make_tuple(newNodeID),
+                        std::make_tuple());
+  directedGraph[newNodeID].emplace(targetID, ConnectionGene(innovID2, 1,
+      newNodeID, targetID));
+
+  // Remove the lost connection.
+  directedGraph[sourceID][targetID].Enabled() = false;
+  connectionGeneList[i].Enabled() = false;
+
+  // If the genome is acyclic, change the depths.
+  if (isAcyclic)
+  {
+    nodeDepths.push_back(nodeDepths[sourceID] + 1);
+
+    // If this is the case, the connection we are splitting is part of the
+    // longest path.
+    if (nodeDepths[targetID] == nodeDepths[sourceID] + 1)
+    {
+      nodeDepths[targetID]++;
+      Traverse(targetID);
+    }
+  }
+}
+
+// Delete connection.
+template <class ActivationFunction>
+void Genome<ActivationFunction>::DelConnMutation()
+{
+  size_t i = 0;
+  i = arma::randi<arma::uvec>(1, arma::distr_param(0,
+      (int)(connectionGeneList.size() - 1)))[0];
+
+  if (connectionGeneList[i].Enabled())
+    return;
+
+  size_t sourceID = connectionGeneList[i].Source();
+  size_t targetID = connectionGeneList[i].Target();
+
+  connectionGeneList[i].Enabled() = false;
+  directedGraph[sourceID][targetID].Enabled() = false;
+
+  // If the genome is acyclic, change the depths.
+  // Think of a better way to do this.
+  if (isAcyclic)
+  {
+    std::fill(nodeDepths.begin(), nodeDepths.end(), 0);
+    for (size_t j = 0; j <= inputNodeCount; j++)
+      Traverse(j);
+  }
+}
+
+// Finds complexity of the genome.
+template <class ActivationFunction>
+size_t Genome<ActivationFunction>::Complexity()
+{
+  size_t connCount = 0;
+  for (size_t i = 0 ; i < connectionGeneList.size(); i++)
+  {
+    if (connectionGeneList[i].Enabled())
+      connCount++;
```

Comment:
I'm wondering if it makes sense to build the connectivity at the same time as we add/disable the connection. Let me know what you think.

Reply:
Do you mean update the nose depths? I tried this, but I couldn't think of a different way to do it.

Reply:
I was thinking we could update the connection count parameter once the genome is created, so if we can a new enabled connection we could increase the counter, if we disable one we could decrease the counter. That way we now at each time what the connection count is. Do you think that is feasible?

Reply:
sounds good to me, I'll make that change

Reply:
Sounds good I think this could take some time.